OpenNefia.Content/UI/Hud/HudLayer.cs

```diff
@@ -1,56 +1,371 @@
+            DateText = new UiText();
+
+            CanKeyboardFocus = true;
+            MessageBoxBacking = new UiMessageWindowBacking();
+            BacklogBacking = new UiMessageWindowBacking(UiMessageWindowBacking.MessageBackingType.Expanded);
+            HudBar = new UiHudBar();
+            MiniMapAsset = Assets.Get(Protos.Asset.HudMinimap);
+            Minimap = new Minimap();
+            ClockAsset = Assets.Get(Protos.Asset.Clock);
+            ClockHand = new ClockHand();
+            DateFrame = Assets.Get(Protos.Asset.DateLabelFrame);
+
+            BarAsset = Assets.Get(Protos.Asset.HpBarFrame);
+            HpBarAsset = Assets.Get(Protos.Asset.HudHpBar);
+            MpBarAsset = Assets.Get(Protos.Asset.HudMpBar);
+
+            HealthText = new UiTextOutlined(UiFonts.HUDBarText);
+            ManaText = new UiTextOutlined(UiFonts.HUDBarText);
+
+            ExpText = new UiTextOutlined(UiFonts.HUDInfoText);
+            GoldText = new UiTextOutlined(UiFonts.HUDInfoText);
+            PlatText = new UiTextOutlined(UiFonts.HUDInfoText);
+            ExpIconAsset = Assets.Get(Protos.Asset.CharacterLevelIcon);
+            GoldIconAsset = Assets.Get(Protos.Asset.GoldCoin);
+            PlatIconAsset = Assets.Get(Protos.Asset.PlatinumCoin);
+
+            SkillIcons = Assets.Get(Protos.Asset.HudSkillIcons);
+            MapNameIcon = Assets.Get(Protos.Asset.MapNameIcon);
+            MapNameText = new UiText(UiFonts.HUDSkillText);
+
+            MessageBoxContainer = new UiVerticalContainer();
+            BacklogContainer = new UiVerticalContainer();
+
+            MessageWindow = new HudMessageWindow(MessageBoxContainer, BacklogContainer);
+
+            foreach (HudSkillIconType type in Enum.GetValues<HudSkillIconType>())
+            {
+                var uiText = new UiText(UiFonts.HUDSkillText);
+                if (type > HudSkillIconType.Cha)
+                    HudSkillContainer.AddLayout(LayoutType.Spacer, 5);
+                HudSkillContainer.AddElement(uiText);
+                SkillTexts[type] = uiText;
+            }
+
+            _field.OnScreenRefresh += OnScreenRefresh;
+            UpdateTime();
+        }
+
+        private void OnScreenRefresh()
+        {
+            UpdateMinimap();
+            UpdateBars();
+            UpdateInfoTexts();
+        }
+
+        public void UpdateTime()
+        {
+            var date = _world.State.GameDate;
+            DateText.Text = $"{date.Year}/{date.Month}/{date.Day}";
+            ClockHand.SetHour(date.Hour);
+        }
+
+        public void UpdateMinimap()
+        {
+            if (_entMan.TryGetComponent<SpatialComponent>(GameSession.Player, out var spatial))
+                Minimap.Refresh(_mapManager.ActiveMap?.TileMemory!, spatial.MapPosition);
+        }
+
+        private void UpdateBars()
+        {
+            if (_entMan.TryGetComponent<SkillsComponent>(GameSession.Player, out var skills))
```

Comment:
Also can the HP/MP bars be their own `UiElement` too?

Reply:
changed in 9eb449f